src/ui/public/state_management/state.js

```diff
@@ -149,6 +181,80 @@ export default function StateProvider(Private, $rootScope, $location) {
     this._defaults = defaults || {};
   };
 
+  /**
+   *  Parse the query param value to it's unserialized
+   *  value. Hashes are restored to their pre-hashed state.
+   *
+   *  @param  {string} queryParam - value from the query string
+   *  @return {any} - the stored value, or null if hash does not resolve
+   */
+  State.prototype._parseQueryParamValue = function (queryParam) {
+    if (!isStateHash(queryParam)) {
+      return rison.decode(queryParam);
+    }
+
+    const json = this._hashedItemStore.getItem(queryParam);
+    if (json === null) {
+      this._notifier.error('Unable to completely restore the URL, be sure to use the share functionality.');
+    }
+
+    return JSON.parse(json);
+  };
+
+  /**
+   *  Lookup the value for a hash and return it's value
+   *  in rison format
+   *
+   *  @param  {string} hash
+   *  @return {string} rison
+   */
+  State.prototype.translateHashToRison = function (hash) {
+    return rison.encode(this._parseQueryParamValue(hash));
+  };
+
+  /**
+   *  Produce the hash version of the state in it's current position
+   *
+   *  @return {string}
+   */
+  State.prototype.toQueryParam = function (state = this.toObject()) {
+    if (!config.get('state:storeInSessionStorage')) {
+      return rison.encode(state);
+    }
+
+    // We need to strip out Angular-specific properties.
+    const json = angular.toJson(state);
+    const hash = createStateHash(json, hash => {
+      return this._hashedItemStore.getItem(hash);
+    });
+    const isItemSet = this._hashedItemStore.setItem(hash, json);
+
+    if (isItemSet) {
+      return hash;
+    }
+
+    // If we ran out of space trying to persist the state, notify the user.
+    this._notifier.fatal(
+      new Error(
+        'Kibana is unable to store history items in your session ' +
+        'because it is full and there don\'t seem to be items any items safe ' +
+        'to delete.\n' +
+        '\n' +
+        'This can usually be fixed by moving to a fresh tab, but could ' +
```

Comment:
Per [Court's comment](https://github.com/elastic/kibana/pull/8022#issuecomment-245089078), let's update to include information for enabling the _state:storeInSessionStorage_ option.